### ISSUE TITLE
Cannot get GPU implementation function: NCCL Failure

### ISSUE THREAD

**Issue by mark-mediware** [dmlc/xgboost#4811]
I have a bug I can't seem to correct. I previously had gpu_hist functional and working, but have been unable to resolve a new issue today.

The example below produces the following error:

```
import xgboost
from sklearn.datasets import load_breast_cancer

param = {'max_depth': 6, 'eta': 1, 'silent': 1,
         'objective': 'binary:logistic',
         'tree_method': 'gpu_hist',
         'n_gpus': -1, 'silent': 0,
         'eta': 0.01, 'eval_metric':'logloss'}

X, y = load_breast_cancer(return_X_y=True)
dtrain = xgboost.DMatrix(X, y)

num_round = 10
bst = xgboost.train(param, dtrain, num_round, [(dtrain, 'train')])
```

`XGBoostError: [01:44:51] /workspace/src/tree/updater_gpu_hist.cu:980: Exception in gpu_hist: NCCL failure :unhandled system error /workspace/src/tree/../common/device_helpers.cuh(952)

Stack trace returned 10 entries:
[bt] (0) /usr/local/xgboost/libxgboost.so(dmlc::StackTrace(unsigned long)+0x47) [0x7fa8e5ecefc7]
[bt] (1) /usr/local/xgboost/libxgboost.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x1d) [0x7fa8e5ecf42d]
[bt] (2) /usr/local/xgboost/libxgboost.so(xgboost::tree::GPUHistMakerSpecialised<xgboost::detail::GradientPairInternal<double> >::Update(xgboost::HostDeviceVector<xgboost::detail::GradientPairInternal<float> >*, xgboost::DMatrix*, std::vector<xgboost::RegTree*, std::allocator<xgboost::RegTree*> > const&)+0x29a) [0x7fa8e615972a]
[bt] (3) /usr/local/xgboost/libxgboost.so(xgboost::gbm::GBTree::BoostNewTrees(xgboost::HostDeviceVector<xgboost::detail::GradientPairInternal<float> >*, xgboost::DMatrix*, int, std::vector<std::unique_ptr<xgboost::RegTree, std::default_delete<xgboost::RegTree> >, std::allocator<std::unique_ptr<xgboost::RegTree, std::default_delete<xgboost::RegTree> > > >*)+0x58f) [0x7fa8e5f60cff]
[bt] (4) /usr/local/xgboost/libxgboost.so(xgboost::gbm::GBTree::DoBoost(xgboost::DMatrix*, xgboost::HostDeviceVector<xgboost::detail::GradientPairInternal<float> >*, xgboost::ObjFunction*)+0x8de) [0x7fa8e5f6210e]
[bt] (5) /usr/local/xgboost/libxgboost.so(xgboost::LearnerImpl::UpdateOneIter(int, xgboost::DMatrix*)+0x3e3) [0x7fa8e5fa82a3]
[bt] (6) /usr/local/xgboost/libxgboost.so(XGBoosterUpdateOneIter+0x35) [0x7fa8e5ec7535]
[bt] (7) /lib64/libffi.so.6(ffi_call_unix64+0x4c) [0x7fa971506dcc]
[bt] (8) /lib64/libffi.so.6(ffi_call+0x1f5) [0x7fa9715066f5]
[bt] (9) /usr/local/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so(_ctypes_callproc+0x283) [0x7fa97171a8f3]`

To date, I've tried re-installing xgboost locally, as well as updated CUDA to v10.1

Looking for next steps here, as a deep error search yielded limited help.

**Comment by trivialfis** [comment]
@mark-mediware Any update?

**Comment by trivialfis** [comment]
Closing as I don't think this is related to XGBoost.

**Comment by trivialfis** [comment]
Feel free to re-open if problem persist.

**Comment by trivialfis** [comment]
In case of anyone who searches this error message 
```
Exception in gpu_hist: NCCL failure :internal error
```

There are multiple reasons nccl can fail:
- Connection failure, sometime this is due to wrong network interface.  I encountered that during running dask.  I resolved it by changing the network interface from `ib<number>` to `bond<number>`.

- Mismatched nccl shared object.  I resolved it by making sure all workers and clients having exact same compiled XGBoost and nccl.
